catatonit.c

```diff
@@ -136,6 +139,50 @@ int kernel_signals[] = {SIGFPE, SIGILL, SIGSEGV, SIGBUS, SIGABRT, SIGTRAP, SIGSY
 
 #define ARRAY_LEN(arr) (sizeof(arr) / sizeof(*arr))
 
+static int syscall_close_range(unsigned int fd, unsigned int max_fd, unsigned int flags)
+{
+  return (int) syscall(__NR_close_range, fd, max_fd, flags);
+}
+
+/*
+ * Close every fd >= n that is different from exclude_fd.
+ */
+static void close_fds_ge_than(int n, int exclude_fd)
+{
+	struct dirent *next;
+	DIR *dir;
+	int ret;
+	int fd;
+
+	ret = syscall_close_range(n, UINT_MAX, 0);
+	if (ret == 0)
+		return;
+
+	dir = opendir("/proc/self/fd");
+	if (dir == NULL) {
+		warn("cannot opendir /proc/self/fd: %m");
+		return;
+	}
+
+	fd = dirfd(dir);
+	for (next = readdir (dir); next; next = readdir (dir)) {
+		int val;
```

Comment:
Might want to use long long for this as that is strotoll returns, to make sure you can fit LLONG_MIN for underflow.